Improve menu item accessibility semantics

flutter-zl opened the pull request:
This PR addresses issue #171231 by enhancing the accessibility semantics for `MenuItemButton` and `SubmenuButton`. Previously, screen readers did not correctly announce the disabled state of these menu items.

This ensures that menu items are correctly identified as buttons and their enabled/disabled state is communicated to assistive technologies.

**Before change**
https://button-0923-before.web.app/

**After change**
https://button-0923-after.web.app/

flutter-zl committed:
Improve menu item accessibility semantics

flutter-zl committed:
Merge branch 'master' into button_0923

flutter-zl committed:
Fix linter error: remove redundant null argument

flutter-zl committed:
Merge branch 'master' into button_0923

flutter-zl committed:
Merge branch 'master' into button_0923

flutter-zl committed:
Merge branch 'master' into button_0923

flutter-zl committed:
Make button semantics for menu items opt-in

mdebbar commented on packages/flutter/lib/src/material/menu_anchor.dart:
```diff
@@ -2893,10 +2945,13 @@ class _MenuItemLabel extends StatelessWidget {
         label: semanticsLabel,
-        excludeSemantics: true,
+        enabled: isEnabled,
+        button: isSemanticButton,
```
I'm not sure that this is the right place to set the semantics `button` property. I think we should be replacing the `isSemanticButton: null` passed to `TextButton` above (line 1886).

Same for `enabled`.

chunhtai replied:
we should programatically assign the button: true based on defaultTargetPlatform, this is not something developer can easily pick and know about the consequence, and it seems  like there may not be a good reason to provide the flexibility

mdebbar reviewed (commented):
@chunhtai @QuncCccccc do you think `isSemanticButton` should be configurable by the user, or just default to the right value for each platform (or both)?

Default value per platform would be:
- Windows - `isSemanticButton: null`
- Web - `isSemanticButton: true`
- Others - ??

chunhtai commented:
I agree we may need to be platform adaptive on this, I think for now Web - `isSemanticButton: true` and null otherwise should be fine

flutter-zl committed:
Make menu item button semantics platform-adaptive

flutter-zl committed:
Remove redundant null check in excludeSemantics

flutter-zl committed:
Improve menu item semantics test clarity and remove redundancy

- Rename semantics tests to clarify they test platform-adaptive behavior
- Remove redundant default semantics tests that would fail on web
- Remove out-of-scope disabled semantics tests to keep PR focused
- Platform-adaptive tests with `if (kIsWeb)` cover all platforms

flutter-zl committed:
revert these accidental changes

mdebbar commented on packages/flutter/lib/src/material/menu_anchor.dart:
```diff
@@ -935,7 +935,7 @@ class _MenuItemButtonState extends State<MenuItemButton> {
       statesController: widget.statesController,
       clipBehavior: widget.clipBehavior,
-      isSemanticButton: null,
+      isSemanticButton: kIsWeb,
```
Is `null` equivalent to `false`?

flutter-zl replied:
good catch. updated.

mdebbar commented on packages/flutter/lib/src/material/menu_anchor.dart:
```diff
@@ -1846,7 +1846,7 @@ class _SubmenuButtonState extends State<SubmenuButton> {
                 onFocusChange: _enabled ? widget.onFocusChange : null,
                 onPressed: _enabled ? toggleShowMenu : null,
-                isSemanticButton: null,
+                isSemanticButton: kIsWeb,
```
Same question here.

flutter-zl replied:
good catch. updated.

mdebbar reviewed (commented):
We changed the value on other platform from `null` to `false`. If those are equivalent, then the PR looks good to me!

flutter-zl committed:
Fix(a11y): Correct MenuItemButton and SubmenuButton semantics
